src/node/coinstats.cpp

```diff
@@ -5,55 +5,79 @@
 
 #include <node/coinstats.h>
 
-#include <coins.h>
+#include <crypto/muhash.h>
 #include <hash.h>
+#include <index/coinstatsindex.h>
 #include <serialize.h>
-#include <validation.h>
 #include <uint256.h>
 #include <util/system.h>
+#include <validation.h>
 
 #include <map>
 
-static void ApplyStats(CCoinsStats &stats, CHashWriter& ss, const uint256& hash, const std::map<uint32_t, Coin>& outputs)
+static void ApplyStats(CCoinsStats &stats, MuHash3072& muhash, const uint256& hash, const std::map<uint32_t, Coin>& outputs)
 {
     assert(!outputs.empty());
-    ss << hash;
-    ss << VARINT(outputs.begin()->second.nHeight * 2 + outputs.begin()->second.fCoinBase ? 1u : 0u);
-    stats.nTransactions++;
     for (const auto& output : outputs) {
-        ss << VARINT(output.first + 1);
-        ss << output.second.out.scriptPubKey;
-        ss << VARINT_MODE(output.second.out.nValue, VarIntMode::NONNEGATIVE_SIGNED);
+        COutPoint outpoint = COutPoint(hash, output.first);
+        Coin coin = output.second;
+
+        muhash *= MuHash3072(GetTruncatedSHA512Hash(outpoint, coin).begin());
+
         stats.nTransactionOutputs++;
         stats.nTotalAmount += output.second.out.nValue;
         stats.nBogoSize += 32 /* txid */ + 4 /* vout index */ + 4 /* height + coinbase */ + 8 /* amount */ +
                            2 /* scriptPubKey len */ + output.second.out.scriptPubKey.size() /* scriptPubKey */;
     }
-    ss << VARINT(0u);
 }
 
+uint256 GetTruncatedSHA512Hash(const COutPoint& outpoint, const Coin& coin) {
+    TruncatedSHA512Writer ss(SER_DISK, 0);
+    ss << outpoint;
+    ss << (uint32_t)(coin.nHeight * 2 + coin.fCoinBase);
+    ss << coin.out;
+    return ss.GetHash();
```

Comment:
Let's document what these fields are and why they're useful:
* `outpoint`: transaction hash + output position: uniquely identifies an unspent output (BIP 30)
* `height` + `coinbase`: determines if it can be spent, e.g. given coinbase maturity and nLockTime 
* `out`: amount + scriptPubKey